src/CodeGen_Hexagon.cpp

```diff
@@ -326,350 +313,386 @@ void CodeGen_Hexagon::compile_func(const LoweredFunc &f,
     CodeGen_Posix::end_func(f.args);
 }
 
+namespace {
+
+struct IdPair {
+    Intrinsic::ID i64 = Intrinsic::not_intrinsic;
+    Intrinsic::ID i128 = Intrinsic::not_intrinsic;
+
+    Intrinsic::ID get(bool is_128B) const { return is_128B ? i128 : i64; }
+};
+
+// LLVM Hexagon HVX intrinsics are broken up into 64B and 128B versions, for example,
+// llvm::Intrinsic::hexagon_V6_vaddh and llvm::Intrinsic::hexagon_V6_vaddh_128B. This
+// macro selects the 64B or 128B mode depending on the value of is_128B.
+// TODO: is there a numerical correlation (e.g. the 128B version can be mathematically
+// derived from the 64B version)? That would allow us to dodge this macro stuff.
+#ifdef WITH_HEXAGON
+#define MAKE_ID_PAIR(i64) IdPair{ (i64), (i64##_128B) }
+#else
+#define MAKE_ID_PAIR(i64) IdPair{ Intrinsic::not_intrinsic, Intrinsic::not_intrinsic }
+#endif
+
 struct HvxIntrinsic {
     enum {
         BroadcastScalarsToWords = 1 << 0,  // Some intrinsics need scalar arguments broadcasted up to 32 bits.
         v65OrLater = 1 << 1,
     };
-    Intrinsic::ID id;
-    Type ret_type;
+    IdPair ids;
+    halide_type_t ret_type;
     const char *name;
-    vector<Type> arg_types;
+    halide_type_t arg_types[4];
     int flags;
 };
 
+constexpr halide_type_t i8 = halide_type_t(halide_type_int, 8);
+constexpr halide_type_t i16 = halide_type_t(halide_type_int, 16);
+constexpr halide_type_t i32 = halide_type_t(halide_type_int, 32);
+constexpr halide_type_t u8 = halide_type_t(halide_type_uint, 8);
+constexpr halide_type_t u16 = halide_type_t(halide_type_uint, 16);
+constexpr halide_type_t u32 = halide_type_t(halide_type_uint, 32);
+
+// Define vectors that are 1x and 2x the Hexagon HVX width --
+// Note that we use placeholders here (which we fix up when processing
+// the table) as we don't know the HVX width until we know the target
+// we're using; this approach lets us make a compact table with static
+// data, rather than having to assemble it at runtime.
+constexpr int kOneX = 64*8;
+
+constexpr halide_type_t i8v1 = i8.with_lanes(kOneX / 8);
+constexpr halide_type_t i16v1 = i16.with_lanes(kOneX / 16);
+constexpr halide_type_t i32v1 = i32.with_lanes(kOneX / 32);
+constexpr halide_type_t u8v1 = u8.with_lanes(kOneX / 8);
+constexpr halide_type_t u16v1 = u16.with_lanes(kOneX / 16);
+constexpr halide_type_t u32v1 = u32.with_lanes(kOneX / 32);
+
+constexpr halide_type_t i8v2 = i8v1.with_lanes(i8v1.lanes * 2);
+constexpr halide_type_t i16v2 = i16v1.with_lanes(i16v1.lanes * 2);
+constexpr halide_type_t i32v2 = i32v1.with_lanes(i32v1.lanes * 2);
+constexpr halide_type_t u8v2 = u8v1.with_lanes(u8v1.lanes * 2);
+constexpr halide_type_t u16v2 = u16v1.with_lanes(u16v1.lanes * 2);
+constexpr halide_type_t u32v2 = u32v1.with_lanes(u32v1.lanes * 2);
+
+// LLVM's HVX vector intrinsics don't include the type of the
+// operands, they all operate on vectors of 32 bit integers. To make
+// it easier to generate code, we define wrapper intrinsics with
```

Comment:
This comment seems partially misplaced, the wrapper intrinsics are still defined in init_module. This table just provides a new name and the arg/return types of the intrinsics.

Reply:
Done.